### PR TITLE
Users importation

### LINKED ISSUE #177: Create the students lists importer
- [ ] Allow to admin to import list of students 
- [ ] Parse students and their level of study from CSV
- [ ] Importe the list in database

### PR DESCRIPTION
**Pull request by FrancoisPog**
closes #177 
- [x] Parse users from a CSV file
- [x] Analyze users list
- [x] Update database to add `us_admin` attribute
- [x] Import users in database
- [x] Create the route to import the csv file
- [x] Docs
- [x] Tests

### CODE REVIEW

File: server/global/users_importation/usersAnalyzer.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ * Test if a login is valid
+ * @param {*} login
+ * @returns {boolean}
+ */
+function isLoginValid(login) {
+  return isString(login) && /^[a-z\d]+$/i.test(login);
+}
+
+/**
+ * Test if a admin property is valid
```
Comment (valentinperignon) [inline server/global/users_importation/usersAnalyzer.js:123]:
```suggestion
 * Test if an admin property is valid
```

**Review by valentinperignon — APPROVED** [review]
LGTMAnsItsAwesome

File: server/controllers/usersImportation.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+  const deleteUploadedFile = () => deleteFiles(filepath).catch(Logger.error);
+
+  const sendServorError = (error, title) => {
+    res.sendServerError(addErrorTitle(error, title));
+    deleteUploadedFile();
+  };
+
+  if (!/\.csv$/i.test(originalname)) {
+    res.sendUsageError(400, "Invalid file format : must be csv ");
```
Comment (nathanaelhoun) [inline server/controllers/usersImportation.js:80]:
```suggestion
    res.sendUsageError(400, "Invalid file format: must be csv ");
```

File: server/global/users_importation/usersImporter.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ */
+export function createSqlToInsertAllUsers(users) {
+  let script = "START TRANSACTION; SET AUTOCOMMIT=0; ";
+
+  script += createSqlToClearRemovedUsers(users);
+
+  script += createSqlToInsertOrUpdateUsers(users);
+
+  script += "COMMIT; SET AUTOCOMMIT=1; ";
+  return script;
```
Comment (nathanaelhoun) [inline server/global/users_importation/usersImporter.js:36]: I'd prefer
```suggestion
  const script = "START TRANSACTION; SET AUTOCOMMIT=0; "
    + createSqlToClearRemovedUsers(users)
    + createSqlToInsertOrUpdateUsers(users)
    + "COMMIT; SET AUTOCOMMIT=1; ";

  return script;
```

File: server/global/users_importation/usersImporter.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+function createSqlToClearRemovedUsers(users) {
+  let logins = users
+    .map(({ login }) => mysql.escape(String(login).substr(0, LOGIN_MAX_LENGTH).toLowerCase()))
+    .join(", ");
+
+  if (logins.length === 0) {
+    logins = ["''"];
+  }
+  const date = mysql.escape(dateFormat(new Date(), "yyyy-mm-dd"));
+  return `  DELETE FROM duel \
```
Comment (nathanaelhoun) [inline server/global/users_importation/usersImporter.js:53]: Are the `\` necessary with the "`" ?

File: server/global/users_importation/usersImporter.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ */
+function createSqlToInsertOrUpdateUsers(users) {
+  const defaultAvatar = mysql.escape(
+    JSON.stringify({
+      eyes: 0,
+      hands: 0,
+      hat: 0,
+      mouth: 0,
+      colorBody: "#0c04fc",
+      colorBG: "#D3D3D3",
```
Comment (nathanaelhoun) [inline server/global/users_importation/usersImporter.js:76]: A comment for the color values could be useful :-)

File: server/global/users_importation/usersImporter.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    })
+  );
+
+  return users
+    .map(({ login, admin }) => {
+      admin = mysql.escape(admin === 1 ? 1 : 0);
+      login = mysql.escape(String(login).substr(0, LOGIN_MAX_LENGTH).toLowerCase());
+      return `INSERT INTO user (us_login,us_admin,us_avatar) \
+              VALUES (${login},${admin},${defaultAvatar}) \
+              ON DUPLICATE KEY UPDATE us_admin = ${admin}, us_deleted = NULL; `;
```
Comment (nathanaelhoun) [inline server/global/users_importation/usersImporter.js:86]: If I understand correctly, this request will not reset the avatar if the user already exists, right?

File: server/test/users_importation/importation_route/files/users.xlsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+LOGIN;ADMIN
```
Comment (nathanaelhoun) [inline server/test/users_importation/importation_route/files/users.xlsx:1]: Why is this an `.xlsx` file?

**Review by nathanaelhoun — APPROVED** [review]
image

**Comment by nathanaelhoun** [comment]
Also it may be good to document how to add the first admin account in the [Production Deployement docs](https://github.com/ValFraNath/glowing-octo-guacamole/wiki/Production-deployment). I propose to do it with an raw sql query for the moment (we might add a more user-friendly way later).

File: server/global/users_importation/usersImporter.js (hunk 4 above)
Comment (FrancoisPog) [reply server/global/users_importation/usersImporter.js:53]: I think no, but it avoid too much spaces

File: server/global/users_importation/usersImporter.js (hunk 6 above)
Comment (FrancoisPog) [reply server/global/users_importation/usersImporter.js:86]: Yep, deleted users keep their stats and avatars.

File: server/test/users_importation/importation_route/files/users.xlsx (hunk 7 above)
Comment (FrancoisPog) [reply server/test/users_importation/importation_route/files/users.xlsx:1]: To test a wrong file format.

File: server/controllers/moleculesImporter.js [hunk 8]
```diff
@@ -2,15 +2,16 @@ import path from "path";
...
 import HttpResponseWrapper from "../global/HttpResponseWrapper.js";
 import { bindImagesToMolecules } from "../global/images_importation/imagesImporter.js";
 import Logger, { addErrorTitle } from "../global/Logger.js";
 import { analyzeData } from "../global/molecules_importation/moleculesAnalyzer.js";
 import { createSqlToInsertAllData } from "../global/molecules_importation/moleculesImporter.js";
 import { parseMoleculesFromCsv } from "../global/molecules_importation/moleculesParser.js";
 
-const FILES_DIR_PATH = path.resolve("files", "molecules");
+const FILES_DIR = process.env.NODE_ENV === "test" ? "files-test" : "files";
+const IMAGES_DIR = path.resolve(FILES_DIR, "molecules");
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/controllers/moleculesImporter.js:14]: Oups I have to `s/IMAGES_DIR/MOLECULES_DIR`

File: server/global/users_importation/usersImporter.js (hunk 5 above)
Comment (FrancoisPog) [reply server/global/users_importation/usersImporter.js:74]: done 👍

**Pull request merged**